compiler/passes/src/const_propagation/const_propagate_expression.rs

```diff
@@ -0,0 +1,335 @@
+
+            let ty = self.type_table.get(&input.array.id());
+            let Some(Type::Array(array_ty)) = ty else {
+                panic!("Type checking guaranteed that this is an array.");
+            };
+            let len = array_ty.length();
+
+            macro_rules! err {
+                ($x: expr) => {
+                    // Only emit a bounds error if we have no other errors yet.
+                    // This prevents a chain of redundant error messages when a loop is unrolled.
+                    if !self.handler.had_errors() {
+                        self.emit_err(StaticAnalyzerError::array_bounds($x, len, span));
+                    }
+                };
+            }
+
+            match value {
+                Value::U8(x) if x as usize >= len => err!(&x),
+                Value::U16(x) if x as usize >= len => err!(&x),
+                Value::U32(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::U64(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::U128(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::I8(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::I16(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::I32(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::I64(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                Value::I128(x) if x.try_into().unwrap_or(len) >= len => err!(&x),
+                _ => {
+                    // Type checking guarantees this is an integer, and none of the above cases matched,
+                    // so we're in bounds and don't need to do anything.
+                }
+            };
+        } else {
+            self.array_index_not_evaluated = Some(index_expr.span());
+        }
+        *input.index = index_expr;
+        (Expression::Access(leo_ast::AccessExpression::Array(input)), None)
+    }
+
+    fn reconstruct_associated_constant(
+        &mut self,
+        input: leo_ast::AssociatedConstant,
+    ) -> (Expression, Self::AdditionalOutput) {
+        // Currently there is only one associated constant.
+        let generator = Value::generator();
+        let expr = value_to_expression(&generator, input.span(), self.node_builder);
+        (expr, Some(generator))
+    }
+
+    fn reconstruct_associated_function(
+        &mut self,
+        mut input: leo_ast::AssociatedFunction,
+    ) -> (Expression, Self::AdditionalOutput) {
+        let mut values = Vec::new();
+        for argument in input.arguments.iter_mut() {
+            let (new_argument, opt_value) = self.reconstruct_expression(std::mem::take(argument));
+            *argument = new_argument;
+            if let Some(value) = opt_value {
+                values.push(value);
+            }
+        }
+
+        if values.len() == input.arguments.len() && !matches!(input.variant.name, sym::CheatCode | sym::Mapping) {
```

Comment:
Should this also check `ChaCha` and `Future`?

Reply:
Ah gotcha! I missed that `CoreFunctionHelper` for Vec<Value>` did not have an rng. 
It could be helpful to have comments on the trait and on the impl to understand the context.
For some reason, I thought we needed tests checking that propagation is not happening on futures and ChaCha, but, you're right. The tests are fine as is.